### PR TITLE
Blog post nullannotations

### PR DESCRIPTION
**Pull request by triller-telekom**
Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

### CODE REVIEW

File: docs/_posts/2017-11-30-nullannotations.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+---
+layout: post
+title: "Less NPE's - We now use Eclipse Nullannotations!"
+date:   2017-11-30
+image: "2017-11-30-nullannotations.png"
+published: true
+---
+
+Nullpointer Exceptions (aka NPE's) are pretty nasty since they are `RuntimeException`s and thus a developers does not immediately notice them while coding in their IDE.
```
Comment (htreu) [inline docs/_posts/2017-11-30-nullannotations.md]: `and thus a developer does` or `and thus developers do`

File: docs/_posts/2017-11-30-nullannotations.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+Hence, data-flow analysis where written that show potential accesses to nullpointers before they occur at runtime. Our project decided to make use of these ananlyses and get rid of (some) nullpointers.
+
+<!--more-->
+
+Two popular projects that offer such an analysis are the [Checker Framework](https://checkerframework.org) and the [Eclipse JDT compiler](https://wiki.eclipse.org/JDT_Core/Null_Analysis).
+In order to be obtain better results from these anylyses, developers can make use of `annotations` in the code.
```
Comment (htreu) [inline docs/_posts/2017-11-30-nullannotations.md]: `In order to obtain better`

File: docs/_posts/2017-11-30-nullannotations.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+In the current snapshots and in our next release we support such annotations in our provided Eclipse IDE setup and maven configuration.
+We decided to use the implementation offered by the Eclipse JDT compiler.
+
+#### Making JavaDoc comments explicit
+
+Since we are a framework that will be integrated into solutions which are build by others, we have provided proper Javadoc documentation on our public interfaces in the past.
+In these Javadoc comments we state that a method will never return `null`, a parameter should never be `null` or is optional and thus maybe `null`.
+Maintaining these comments by hand is error-prone and subject to be ignored by developers not reading them.
+Therefore we have decided to provide these comments in a machine readable way so the IDE can support the developers in showing them errors or warnings in case they violate them.
```
Comment (htreu) [inline docs/_posts/2017-11-30-nullannotations.md]: `so the IDE can support developers in` (`the` is obsolete here)

File: docs/_posts/2017-11-30-nullannotations.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+If a specific field, return value, or parameter in the class should be allowed to become `null` we annotate them with `@Nullable`.
+Basically what these annotations are doing for us is translating the contract written in the Javadoc comment into something that can automatically be checked by the compiler.
+
+We have already started to annotate our core packages so binding developers can benefit from them.
+
+### Real life examples
+
+One simple example is the [ThingHandler:handleCommand(ChannelUID channelUID, Command command)](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingHandler.java#L99) which guarantees the binding developers that both arguments are **not** `null` and this they can simply implement their logic by going through the IDs via `switch (channelUID.getId())` without running into a NPE.
```
Comment (htreu) [inline docs/_posts/2017-11-30-nullannotations.md]: `which guarantees to binding developers`

Comment (htreu) [reply docs/_posts/2017-11-30-nullannotations.md]: `**not** `null` so they can simply implement their`

File: docs/_posts/2017-11-30-nullannotations.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+{:.center}
+![NonNull annotation]({{ "/img/blog/2017-11-30-nullannotations_handleCommand.png" | absolute_url }})
+
+For bindings that are using bridge things and call [BaseThingHandler:getBridge()](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L584) we now have the `@Nullable` annotation in place that reminds the developers to do a proper `null` check before using the return value of this method.
```
Comment (htreu) [inline docs/_posts/2017-11-30-nullannotations.md]: could also be `For bindings which use bridge things and call`

Comment (htreu) [reply docs/_posts/2017-11-30-nullannotations.md]: `that reminds developers to do`

File: docs/_posts/2017-11-30-nullannotations.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+{:.center}
+![Nullable annotation with warning]({{ "/img/blog/2017-11-30-nullannotations_bridgeNullable.png" | absolute_url }})
+
+After the `null` check has been implemented the warning is gone:
+
+{:.center}
+![Nullable annotation without warning]({{ "/img/blog/2017-11-30-nullannotations_bridgeNullableCorrected.png" | absolute_url }})
+
+If the developers violate the `null` specifications, the IDE will immediately show an error.
```
Comment (htreu) [inline docs/_posts/2017-11-30-nullannotations.md]: `If developers violate`

File: docs/_posts/2017-11-30-nullannotations.md [hunk 7]
```diff
@@ -6,13 +6,13 @@ image: "2017-11-30-nullannotations.png"
...
 
-Nullpointer Exceptions (aka NPE's) are pretty nasty since they are `RuntimeException`s and thus a developers does not immediately notice them while coding in their IDE.
+Nullpointer Exceptions (aka NPE's) are pretty nasty since they are `RuntimeException`s and thus developers do not immediately notice them while coding in their IDE.
 Hence, data-flow analysis where written that show potential accesses to nullpointers before they occur at runtime. Our project decided to make use of these ananlyses and get rid of (some) nullpointers.
 
 <!--more-->
 
 Two popular projects that offer such an analysis are the [Checker Framework](https://checkerframework.org) and the [Eclipse JDT compiler](https://wiki.eclipse.org/JDT_Core/Null_Analysis).
-In order to be obtain better results from these anylyses, developers can make use of `annotations` in the code.
+In order to obtain better results from these anylyses, developers can make use of `annotations` in the code.
```
Comment (htreu) [inline docs/_posts/2017-11-30-nullannotations.md]: `analysis`

Comment (triller-telekom) [reply docs/_posts/2017-11-30-nullannotations.md]: Its analyses (plural) see: https://www.dict.cc/?s=Analysen

But nevertheless I had a typo in there :P

File: docs/_posts/2017-11-30-nullannotations.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+Two popular projects that offer such an analysis are the [Checker Framework](https://checkerframework.org) and the [Eclipse JDT compiler](https://wiki.eclipse.org/JDT_Core/Null_Analysis).
+In order to obtain better results from these anylyses, developers can make use of `annotations` in the code.
+In the current snapshots and in our next release we support such annotations in our provided Eclipse IDE setup and maven configuration.
+We decided to use the implementation offered by the Eclipse JDT compiler.
+
+#### Making JavaDoc comments explicit
+
+Since we are a framework that will be integrated into solutions which are build by others, we have provided proper Javadoc documentation on our public interfaces in the past.
+In these Javadoc comments we state that a method will never return `null`, a parameter should never be `null` or is optional and thus maybe `null`.
+Maintaining these comments by hand is error-prone and subject to be ignored by developers not reading them.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: > "subject to be ignored"

"easily overseen"? _Ignoring_ something is done deliberately...

File: docs/_posts/2017-11-30-nullannotations.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+Therefore we have decided to provide these comments in a machine readable way so the IDE can support developers in showing them errors or warnings in case they violate them.
+
+As described in our [Coding Guidelines](https://www.eclipse.org/smarthome/documentation/development/guidelines.html#a-code-style) we annotate classes with the `@NonNullByDefault` annotations which declares that every field, return value, parameter, etc. defined in this class will always have a value unequal to `null`.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: * As described in our [Coding Guidelines](https://www.eclipse.org/smarthome/documentation/development/guidelines.html#a-code-style)**,** [...]

* What is "etc."? Isn't this list complete already?

* "defined in this class" -> "defined in such classes"

* "will always have a value unequal to `null`" "will always have value**s** other than `null`" (or "...not equals to `null`)

File: docs/_posts/2017-11-30-nullannotations.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+If a specific field, return value, or parameter in the class should be allowed to become `null` we annotate them with `@Nullable`.
+Basically what these annotations are doing for us is translating the contract written in the Javadoc comment into something that can automatically be checked by the compiler.
+
+We have already started to annotate our core packages so binding developers can benefit from them.
+
+### Real life examples
+
+One simple example is the [ThingHandler:handleCommand(ChannelUID channelUID, Command command)](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingHandler.java#L99) which guarantees to binding developers that both arguments are **not** `null` and so they can simply implement their logic by going through the IDs via `switch (channelUID.getId())` without running into a NPE.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: To make it less confusing, simply break it up into shorter sentences like this:

```md
One simple example is the [ThingHandler:handleCommand(ChannelUID channelUID, Command command)](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingHandler.java#L99). 
It guarantees to binding developers that both arguments are **not** `null`. 
Therefore they can safely implement their logic by going through the IDs via `switch (channelUID.getId())` without the risk of running into a NPE.
```

File: docs/_posts/2017-11-30-nullannotations.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+{:.center}
+![NonNull annotation]({{ "/img/blog/2017-11-30-nullannotations_handleCommand.png" | absolute_url }})
+
+For bindings which use bridge things and call [BaseThingHandler:getBridge()](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L584) we now have the `@Nullable` annotation in place that reminds developers to do a proper `null` check before using the return value of this method.
+If you omit the `null` check, developers will get a warning as shown below.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: Without the `null` check, implementors [...]:

File: docs/_posts/2017-11-30-nullannotations.md [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+{:.center}
+![Nullable annotation with warning]({{ "/img/blog/2017-11-30-nullannotations_bridgeNullable.png" | absolute_url }})
+
+After the `null` check has been implemented the warning is gone:
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: After the `null` check has been implemented **,** the warning is gone:

File: docs/_posts/2017-11-30-nullannotations.md [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+{:.center}
+![Nullable annotation without warning]({{ "/img/blog/2017-11-30-nullannotations_bridgeNullableCorrected.png" | absolute_url }})
+
+If developers violate the `null` specifications, the IDE will immediately show an error.
+In the example below the `BaseThingHandler:updateState(String channelID, State state)` method requires both parameters to be unequal to `null`.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: * In the example below **,** [...]

* unequal to `null` -> non-`null`

File: docs/_posts/2017-11-30-nullannotations.md [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+But since in the `else` branch the variable used as a second parameter in this method call is set to `null`, the IDE infers that `null` is passed as an argument to the method and shows an error.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: ~~But~~ Since [...]

Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: "second parameter in this method call" -> "second argument"

File: docs/_posts/2017-11-30-nullannotations.md [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+{:.center}
+![Null violation]({{ "/img/blog/2017-11-30-nullannotations_error.png" | absolute_url }})
+
+If you are interested to see a fully annotated binding, please have a look at the [Philips Hue binding](https://github.com/eclipse/smarthome/tree/master/extensions/binding/org.eclipse.smarthome.binding.hue).
+
+
+### Outlook: Third party code
+
+If we make use of external libraries and use their return values in our code the compiler cannot decide whether these values are always unequal to `null` or might become `null` at some point.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: missing comma in the middle or break it up into two sentences

File: docs/_posts/2017-11-30-nullannotations.md [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+This will generate a warning that is currently switched off by our setup.
+In the future we are planing to support *external annotations* which are files that contain information about the method signatures of external libraries.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: Comma: In the future, [...]

File: docs/_posts/2017-11-30-nullannotations.md [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+The [Last NPE project](http://last-npe.org) has started a crowd-sourcing initative to create these external annotations and we are planing to contribute our work to this project.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: Err, that's nice of you... But isn't it rather the plan to _use_ this project's deliverables?

File: docs/_posts/2017-11-30-nullannotations.md [hunk 18]
```diff
@@ -0,0 +1,65 @@
+---
+layout: post
+title: "Less NPE's - We now use Eclipse Nullannotations!"
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: "We" and "Nullannotations" (see my explanations for both somewhere below)

File: docs/_posts/2017-11-30-nullannotations.md [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+
+Two popular projects that offer such an analysis are the [Checker Framework](https://checkerframework.org) and the [Eclipse JDT compiler](https://wiki.eclipse.org/JDT_Core/Null_Analysis).
+In order to obtain better results from these analyses, developers can make use of `annotations` in the code.
+In the current snapshots and in our next release we support such annotations in our provided Eclipse IDE setup and maven configuration.
+We decided to use the implementation offered by the Eclipse JDT compiler.
+
+#### Making JavaDoc comments explicit
+
+Since we are a framework that will be integrated into solutions which are build by others, we have provided proper Javadoc documentation on our public interfaces in the past.
+In these Javadoc comments we state that a method will never return `null`, a parameter should never be `null` or is optional and thus maybe `null`.
```
Comment (sjka) [inline docs/_posts/2017-11-30-nullannotations.md]: maybe -> may be

**Pull request merged**